subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/junit/RerunPreviousFailedTestIntegrationTest.groovy

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.testing.junit
+
+import org.gradle.integtests.fixtures.DefaultTestExecutionResult
+import org.gradle.integtests.fixtures.TargetCoverage
+import org.gradle.testing.fixture.JUnitMultiVersionIntegrationSpec
+import org.hamcrest.Matchers
+import spock.lang.Unroll
+
+import static org.gradle.testing.fixture.JUnitCoverage.JUNIT_4_LATEST
+import static org.gradle.testing.fixture.JUnitCoverage.JUNIT_VINTAGE_JUPITER
+
+@TargetCoverage({ JUNIT_4_LATEST + JUNIT_VINTAGE_JUPITER })
+class RerunPreviousFailedTestIntegrationTest extends JUnitMultiVersionIntegrationSpec {
+    static final String INDEX_OF_TEST_TO_FAIL = "index.of.test.to.fail"
+
+    def setup() {
+        buildFile << """
+            apply plugin: 'java'
+
+            ${mavenCentralRepository()}
+
+
+            dependencies {
+                testCompile '$dependencyNotation'
+            }
+
+            tasks.withType(Test) {
+                maxParallelForks = 1 
+            }
+        """
+
+        (1..3).each {
+            file("src/test/java/ConditionalFailingTest_${it}.java") << """
+                import org.junit.Test;
+                public class ConditionalFailingTest_${it} {
+                    @Test
+                    public void failedTest() {
+                        if("${it}".equals(System.getProperty("${INDEX_OF_TEST_TO_FAIL}"))) {
+                            throw new RuntimeException();
+                        }
+                    }
+                }
+            """.stripIndent()
+        }
+    }
+
+    def letTestFail(def index) {
+        buildFile << """
+        test {
+            systemProperty('${INDEX_OF_TEST_TO_FAIL}', '${index}')
+        }
+        """
+    }
+
+    def 'this is a guard method'(){
```

Comment:
There's a potential bug in our test infrastrucure: without this "guard method`, `gradle testingJvm:iT --tests '* RerunPreviousFailedTestIntegrationTest*'` will complain about no tests found. Not sure the root cause, it needs deep investigation.